c/tskit/tables.c

```diff
@@ -4954,13 +4954,132 @@ tsk_table_sorter_sort_mutations_canonical(tsk_table_sorter_t *self)
+        incoming_edge_count[i] = 0;
+        individuals_todo[i] = TSK_NULL;
+        new_id_map[i] = TSK_NULL;
+    }
+    individuals_todo[num_individuals] = TSK_NULL; /* Sentinel value */
+
+    /* First find the set of individuals that have no children by creating
+     * an array of incoming edge counts */
+    for (i = 0; i < (tsk_id_t) individuals->parents_length; i++) {
+        if (individuals->parents[i] != TSK_NULL) {
+            incoming_edge_count[individuals->parents[i]] += 1;
+        }
+    }
+    /* Use these as the starting points for checking all individuals,
+     * doing this in reverse makes the sort stable */
+    for (i = (tsk_id_t) num_individuals - 1; i >= 0; i--) {
+        if (incoming_edge_count[i] == 0) {
+            individuals_todo[todo_insertion_point] = i;
+            todo_insertion_point++;
+        }
+    }
+
+    /* Now emit individuals from the set that have no children, removing their edges
+     * as we go adding new individuals to the no children set. */
+    while (individuals_todo[current_todo] != TSK_NULL) {
+        tsk_individual_table_get_row_unsafe(
+            individuals, individuals_todo[current_todo], &individual);
+        for (i = 0; i < (tsk_id_t) individual.parents_length; i++) {
+            if (individual.parents[i] != TSK_NULL) {
+                incoming_edge_count[individual.parents[i]]--;
+                if (incoming_edge_count[individual.parents[i]] == 0) {
+                    individuals_todo[todo_insertion_point] = individual.parents[i];
+                    todo_insertion_point++;
+                }
+            }
+        }
+        current_todo++;
+    }
+
+    /* Any edges left are parts of cycles */
+    for (i = 0; i < (tsk_id_t) num_individuals; i++) {
+        if (incoming_edge_count[i] > 0) {
+            ret = TSK_ERR_INDIVIDUAL_PARENT_CYCLE;
+            goto out;
+        }
+    }
+
+    ret = tsk_individual_table_clear(individuals);
+    if (ret != 0) {
+        goto out;
+    }
+
+    /* The sorted individuals are in reverse order */
+    for (i = (tsk_id_t) num_individuals - 1; i >= 0; i--) {
+        tsk_individual_table_get_row_unsafe(
+            &copy, individuals_todo[i], &copy_individual);
+        ret = tsk_individual_table_add_row(individuals, copy_individual.flags,
+            copy_individual.location, copy_individual.location_length,
+            copy_individual.parents, copy_individual.parents_length,
+            copy_individual.metadata, copy_individual.metadata_length);
+        if (ret < 0) {
+            goto out;
+        }
+        new_id_map[individuals_todo[i]] = ret;
+    }
+
+    /* Rewrite the parent ids */
+    for (i = 0; i < (tsk_id_t) individuals->parents_length; i++) {
+        if (individuals->parents[i] != TSK_NULL) {
+            individuals->parents[i] = new_id_map[individuals->parents[i]];
+        }
+    }
+    /* Rewrite the node individual ids */
+    for (i = 0; i < (tsk_id_t) self->tables->nodes.num_rows; i++) {
```

Comment:
Might be worth taking a reference to the node table above (like we do for the individual table) to avoid a bit of pointer chasing in this loop.

Reply:
Done